### PR TITLE
Offline Mode Support

### PR DESCRIPTION
**Pull request by MrFlashAccount**
#### Tl;dr 
Closes: enso-org/cloud-v2#1283
This PR significantly reimplements Offline mode

<details><summary>Demo Presentation</summary>
<p>

https://github.com/enso-org/enso/assets/61194245/752d0423-9c0a-43ba-91e3-4a6688f77034

</p>
</details> 

---

#### Context:
Offline mode is one of the core features of the dashboard. Unfortunately, after adding new features and a few refactoring,  we lost the ability to work offline. 
This PR should bring this functionality back, with a few key differences:
1. We require users to sign in before using the dashboard even in local mode.
2. Once a user is logged in, we allow him to work with local files
3. If a user closes the dashboard, and then open it, he can continue using it in offline mode 

#### This Change:
What does this change do in the larger context? Specific details to highlight for review:
1. Reimplements `<AuthProvider />` functionality, now it implemented on top of `<Suspense />` and ReactQuery
2. Reimplements Backend module flow, now remote backend is always created, You no longer need to check if the RemoteBackend is present
3. Introduces new `<Suspense />` component, which is aware of offline status
4. Introduce new offline-related hooks
5. Add a banner to the form if it's unable to submit it offline
6. Refactor `InviteUserDialog` to the new `<Form />` component
7. Fixes redirect bug when the app doesn't redirect a user to the dashboard after logging in
8. Fixes strange behavior when `/users/me` could stuck into infinite refetch
9. Redesign the Cloud table for offline mode.
10. Adds blocking UI dialog when a user clicks "log out" button

#### Test Plan:
This PR requires thorough QA on the login flow across the browser and IDE. All redirect logic must stay unchanged. 

---

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.

### CODE REVIEW

File: app/ide-desktop/lib/dashboard/src/providers/AuthProvider.tsx [hunk 1]
```diff
@@ -580,28 +437,31 @@ export default function AuthProvider(props: AuthProviderProps) {
       gtagEvent('cloud_sign_out')
       cognito.saveAccessToken(null)
       localStorage.clearUserSpecificEntries()
-      setInitialized(false)
       sentry.setUser(null)
-      setUserSession(null)
+
+      // setUserSession(null)
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/providers/AuthProvider.tsx]: commented code

File: app/ide-desktop/lib/dashboard/src/providers/AuthProvider.tsx [hunk 2]
```diff
@@ -580,28 +437,31 @@ export default function AuthProvider(props: AuthProviderProps) {
...
       // If the User Menu is still visible, it breaks when `userSession` is set to `null`.
       unsetModal()
       // This should not omit success and error toasts as it is not possible
       // to render this optimistically.
-      await toast.toast.promise(cognito.signOut(), {
-        success: getText('signOutSuccess'),
-        error: getText('signOutError'),
-        pending: getText('loggingOut'),
-      })
+      await toast.toast.promise(
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/providers/AuthProvider.tsx]: i guess this is what we want to replace with a fullscreen uncloseable dialog

**Review by somebody1234 — COMMENTED** [review]
i guess the current plan is to fix the logout issue in this PR instead?

i guess it depends on when this PR is expected to be ready

File: app/ide-desktop/lib/dashboard/src/App.tsx [hunk 3]
```diff
@@ -255,19 +272,19 @@ function AppRouter(props: AppRouterProps) {
...
-        window.removeEventListener('beforeunload', logCloseEvent)
-        logCloseEvent()
-      }
-    } else {
-      return
-    }
-  }, [remoteBackend])
+  // React.useEffect(() => {
+  //   if (remoteBackend) {
+  //     void remoteBackend.logEvent('open_app')
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/App.tsx]: commented code, note that we want this to still be here though. perhaps this should be conditionally run only when not in dev mode?

File: app/ide-desktop/lib/dashboard/src/components/AriaComponents/Form/Form.tsx [hunk 4]
```diff
@@ -114,6 +116,15 @@ export const Form = React.forwardRef(function Form<
   // There is no way to avoid type casting here
   // eslint-disable-next-line @typescript-eslint/no-explicit-any,no-restricted-syntax,@typescript-eslint/no-unsafe-argument
   const formOnSubmit = innerForm.handleSubmit(formMutation.mutateAsync as any)
+
+  offlineHooks.useOfflineChange(isOffline => {
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/components/AriaComponents/Form/Form.tsx]: note that some dialogs may need to work on the Local Backend - for example, dialogs for creating e.g. a secret (although this is currently not possible)

File: app/ide-desktop/lib/dashboard/src/providers/AuthProvider.tsx (hunk 2 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/providers/AuthProvider.tsx]: yeah

File: app/ide-desktop/lib/dashboard/src/App.tsx (hunk 3 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/App.tsx]: Yes, I remember, this is why this PR is in draft 😉

File: app/ide-desktop/lib/dashboard/src/components/AriaComponents/Form/Form.tsx (hunk 4 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/components/AriaComponents/Form/Form.tsx]: Good catch! Will add an option to disable this behavior

File: app/ide-desktop/lib/dashboard/src/pages/authentication/SetUsername.tsx [hunk 5]
```diff
@@ -26,13 +26,16 @@ export default function SetUsername() {
...
+  const localBackend = backendProvider.useLocalBackend()
+  const supportsOffline = localBackend != null
 
   const [username, setUsername] = React.useState('')
 
   return (
     <AuthenticationPage
       data-testid="set-username-panel"
       title={getText('setYourUsername')}
+      supportsOffline={supportsOffline}
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/pages/authentication/SetUsername.tsx:38]: might need docs (in `AuthenticationPage.tsx`) for this prop - e.g. if it is always `localBackend != null`, is there a reason it needs to be on every page?

if not, what does it mean when this is accessible when the local backend is available?

File: app/ide-desktop/lib/dashboard/src/layouts/AssetsTable.tsx [hunk 6]
```diff
@@ -1496,7 +1496,7 @@ export default function AssetsTable(props: AssetsTableProps) {
           projectState: {
             type: backendModule.ProjectState.placeholder,
             volumeId: '',
-            ...(user != null ? { openedBy: user.email } : {}),
+            ...{ openedBy: user.email },
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/layouts/AssetsTable.tsx]: the `...{` is no longer necessary

File: app/ide-desktop/lib/dashboard/src/layouts/AssetsTable.tsx [hunk 7]
```diff
@@ -1713,7 +1713,7 @@ export default function AssetsTable(props: AssetsTableProps) {
...
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/layouts/AssetsTable.tsx]: (the `...{` is no longer necessary here as well)

**Review by somebody1234 — APPROVED** [review]

File: app/ide-desktop/lib/dashboard/src/pages/authentication/SetUsername.tsx (hunk 5 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/pages/authentication/SetUsername.tsx:38]: Technically, there is no reason 🙂, but I see AuthenticationPage more like a presentational component, so it's not his responsibility to know what `supportsOffline` means.

Ideally, we need something like a `<AppContext />` that stores: 

1. env vars
2. App level related parameters, like `supportLocalBackend`

Then we can consume this logic across the app and lower-level components will be unaware about this

**Comment by MrFlashAccount** [comment]
> i guess the current plan is to fix the logout issue in this PR instead?
i guess it depends on when this PR is expected to be ready

Yeah, I added a Dialog in this PR.
@PabloBuchu would you like to take a look?

**Comment by PabloBuchu** [comment]
1. Starting local projects after I disconnect from network doesnt work! It comes back after I reconnect.

https://github.com/enso-org/enso/assets/2855109/5506db0c-4b93-4fca-b0a1-5b1c948104db



2. Reopening in offline shows empty cloud dashboard with bunch of errors "Failed to fetch". It should alaways land on local view
Screenshot 2024-06-20 at 10 16 13


3. When I run `npm run dev` in app/gui2 then disconnect I still can browser cloud / recent / trash with errors

Screenshot 2024-06-20 at 10 20 14

**Comment by MrFlashAccount** [comment]
This is not how it is supposed to work in my head. Maybe I broke something during the refactoring? Will take a look!

**Comment by PabloBuchu** [comment]
Two major issues:

1. Running local project after disconnecting is impossible (although all comes back after connecting back to the network)

https://github.com/enso-org/enso/assets/2855109/a578e568-6065-4151-bd7d-a7731bacda30

2. When the credentials are there I should be able to use Enso with local only backend
Screenshot 2024-06-20 at 16 09 47
Screenshot 2024-06-20 at 16 09 50

**Comment by MrFlashAccount** [comment]
@PabloBuchu Hi! fixed both bugs.

**Review by PabloBuchu — APPROVED** [review]
